Data/Aeson/Types/FromJSON.hs

```diff
@@ -1273,38 +1274,71 @@ instance (ProductFromJSON arity f, ProductSize f
 
 --------------------------------------------------------------------------------
 
+class FieldNames f where
+    fieldNames :: f a -> [Text]
+
+instance (FieldNames a, FieldNames b) => FieldNames (a :*: b) where
+    fieldNames _ =
+      fieldNames (undefined :: a x) ++
+      fieldNames (undefined :: b y)
+
+instance (Selector s) => FieldNames (S1 s f) where
+    fieldNames _ = [pack $ selName (undefined :: M1 _i s _f _p)]
+
 class RecordFromJSON arity f where
     recordParseJSON
+        :: Bool :* ConName :* TypeName :* Options :* FromArgs arity a
+        -> Object -> Parser (f a)
+
+instance ( FieldNames f
+         , RecordFromJSON' arity f
+         ) => RecordFromJSON arity f where
+    recordParseJSON (fromTaggedSum :* p@(_ :* _ :* opts :* _)) obj =
+        checkUnknown >> recordParseJSON' p obj
+        where
+            knownFields = (if fromTaggedSum
+                           then (pack (tagFieldName (sumEncoding opts)) :)
+                           else id) $
+                fieldNames (undefined :: f a)
+            checkUnknown =
+                if not (rejectUnknownFields opts)
+                then return ()
+                else case H.keys obj \\ knownFields of
+                    [] -> return ()
+                    unknownFields -> fail ("Unknown fields: " ++ show unknownFields)
```

Comment:
For a more uniform-looking error message, make this start with lower case and (generics only) append some context with `contextCons`.

```suggestion
                    unknownFields -> contextCons cname tname (fail ("unknown fields: " ++ show unknownFields))
```

`cname` and `tname` would be the two arguments before `opts` inside `p`.

It would be good to add a test case for this to `test/ErrorMessages.hs` (and then run the test with `--accept` to refresh the reference messages).